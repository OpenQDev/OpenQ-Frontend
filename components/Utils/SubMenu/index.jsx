import React from 'react';

const SubMenu = ({ updatePage, internalMenu, items, styles, colour, vertical }) => {
  const menu = internalMenu == 'Fixed Contest' ? 'Contest' : internalMenu;
  return (
    <div
      className={`px-2 ${!vertical ? 'sm:px-8' : 'px-0'} text-primary  w-full flex custom-scrollbar  ${
        vertical ? 'flex-col gap-2' : 'h-12 overflow-x-auto overflow-y-hidden border-web-gray border-b items-center'
      }  gap-x-1 md:gap-x-2 lg:gap-x-3 relative ${styles} `}
    >
      {items.map(
        (item, index) =>
          Object.keys(item).length > 0 && (
            <li key={index} className='w-fit list-none  relative'>
              <button
                type='button'
                onClick={() => updatePage(item.name)}
                className={`${
                  vertical && 'pl-2'
                } cursor-pointer px-1 flex gap-1 sm:gap-2 items-center text-sm hover:bg-inactive-gray leading-5 py-1 hover:bg-active-gray rounded-sm justify-center w-fit ${
                  menu === item.name &&
                  ` ${vertical && 'after:h-5 after:left-0 after:w-0.5 after:absolute after:bg-link-colour'} `
                }`}
              >
                {item.Svg && <item.Svg />}
                <span className='whitespace-nowrap'>{item.name}</span>
                {item.SecondSvg && <item.SecondSvg />}
              </button>
              {!vertical && (
                <div
                  className={`absolute w-full  top-9 border ${
                    menu === item.name
                      ? colour === 'rust'
                        ? 'border-rust bg-rust'
                        : 'border-link-colour bg-link-colour'
                      : 'border-transparent'
                  }`}
                ></div>
              )}
            </li>
          )
      )}
    </div>
  );
};

export default SubMenu;
