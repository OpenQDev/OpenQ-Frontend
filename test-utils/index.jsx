// Third party
import React from 'react';
// Cusotm
import { render, renderHook } from '@testing-library/react';
import { cleanup } from '@testing-library/react';

import StoreProvider from '../store/Store/TestStoreProvider';
import AuthProvider from '../store/AuthStore/TestAuthProvider';
import FundProvider from '../components/FundBounty/FundStore/TestFundProvider';
import TokenProvider from '../components/TokenSelection/TokenStore/TokenProvider';

// Add in any providers here if necessary:
// test-utils.js
// (ReduxProvider, ThemeProvider, etc)
afterEach(() => {
  cleanup();
});

const customRender = (ui, options = {}, storeProps, authProps) => {
  const IntersectionObserverMock = vi.fn(() => ({
    disconnect: vi.fn(),
    observe: vi.fn(),
    takeRecords: vi.fn(),
    unobserve: vi.fn(),
  }));
  vi.stubGlobal('IntersectionObserver', IntersectionObserverMock);
  const Providers = ({ children }) => {
    return (
      <AuthProvider StoreProps={authProps}>
        <StoreProvider StoreProps={storeProps}>
          <TokenProvider>{children}</TokenProvider>
        </StoreProvider>
      </AuthProvider>
    );
  };

  return render(ui, { wrapper: Providers, ...options });
};
const customHookRender = (hook, options = {}, storeProps, authProps, fundProps) => {
  const Providers = ({ children }) => {
    return (
      <AuthProvider StoreProps={authProps}>
        <StoreProvider StoreProps={storeProps}>
          <FundProvider StoreProps={fundProps}>
            <TokenProvider>{children}</TokenProvider>
          </FundProvider>
        </StoreProvider>
      </AuthProvider>
    );
  };

  return renderHook(hook, { wrapper: Providers, ...options });
};

// re-export everything
export * from '@testing-library/react';

// override render method
export { customRender as render, customHookRender as hookRender };
