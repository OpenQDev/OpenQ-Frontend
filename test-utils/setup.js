import { expect, beforeEach } from 'vitest';
import { cleanup } from '@testing-library/react';
import matchers from '@testing-library/jest-dom/matchers';

expect.extend(matchers);
beforeEach(() => {
  cleanup();
});
process.on('unhandledRejection', (reason) => {
  // eslint-disable-next-line no-console
  console.log(`FAILED TO HANDLE PROMISE REJECTION`);
  throw reason;
});
