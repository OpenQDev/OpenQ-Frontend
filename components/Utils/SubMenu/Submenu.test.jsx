/**
 * @vi-environment jsdom
 */
import React from 'react';
import { render, screen } from '../../../test-utils';
import SubMenu from '.';
import userEvent from '@testing-library/user-event';
import Add from '../../svg/add';
import renderer from 'react-test-renderer';
// Test cases for full balances, empty balances, and undefined balances.

describe('SubMenu', () => {
  // Test cases for
  const mockUpdatePage = vi.fn();
  beforeEach(() => {});
  const items = [
    { name: 'View', SVG: Add },
    { name: 'Fund', SVG: Add },
  ];
  it('should match DOM Snapshot and match horizontal style', () => {
    const tree = renderer.create(<SubMenu updatePage={mockUpdatePage} internalMenu={'Fund'} items={items} />);
    expect(tree.toJSON()).toMatchSnapshot();
  });
  it('should match DOM Snapshot and match vertical style', () => {
    const tree = renderer.create(
      <SubMenu updatePage={mockUpdatePage} vertical={true} internalMenu={'Fund'} items={items} />
    );
    expect(tree.toJSON()).toMatchSnapshot();
  });
  it('should display SubMenu', async () => {
    // ARRANGE
    const user = userEvent.setup();
    render(<SubMenu updatePage={mockUpdatePage} internalMenu={'Fund'} items={items} />);
    // ASSERT
    const view = screen.getByText(items[0].name);
    await user.click(view);
    expect(mockUpdatePage).toHaveBeenCalled();
    expect(view).toBeInTheDocument();
  });
});
