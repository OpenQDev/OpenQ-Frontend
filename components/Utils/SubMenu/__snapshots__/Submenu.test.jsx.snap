// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SubMenu > should match DOM Snapshot and match horizontal style 1`] = `
<div
  className="px-2 sm:px-8 text-primary  w-full flex custom-scrollbar  h-12 overflow-x-auto overflow-y-hidden border-web-gray border-b items-center  gap-x-1 md:gap-x-2 lg:gap-x-3 relative undefined "
>
  <li
    className="w-fit list-none  relative"
  >
    <button
      className="undefined cursor-pointer px-1 flex gap-1 sm:gap-2 items-center text-sm hover:bg-inactive-gray leading-5 py-1 hover:bg-active-gray rounded-sm justify-center w-fit false"
      onClick={[Function]}
      type="button"
    >
      <span
        className="whitespace-nowrap"
      >
        View
      </span>
    </button>
    <div
      className="absolute w-full  top-9 border border-transparent"
    />
  </li>
  <li
    className="w-fit list-none  relative"
  >
    <button
      className="undefined cursor-pointer px-1 flex gap-1 sm:gap-2 items-center text-sm hover:bg-inactive-gray leading-5 py-1 hover:bg-active-gray rounded-sm justify-center w-fit  undefined "
      onClick={[Function]}
      type="button"
    >
      <span
        className="whitespace-nowrap"
      >
        Fund
      </span>
    </button>
    <div
      className="absolute w-full  top-9 border border-link-colour bg-link-colour"
    />
  </li>
</div>
`;

exports[`SubMenu > should match DOM Snapshot and match vertical style 1`] = `
<div
  className="px-2 px-0 text-primary  w-full flex custom-scrollbar  flex-col gap-2  gap-x-1 md:gap-x-2 lg:gap-x-3 relative undefined "
>
  <li
    className="w-fit list-none  relative"
  >
    <button
      className="pl-2 cursor-pointer px-1 flex gap-1 sm:gap-2 items-center text-sm hover:bg-inactive-gray leading-5 py-1 hover:bg-active-gray rounded-sm justify-center w-fit false"
      onClick={[Function]}
      type="button"
    >
      <span
        className="whitespace-nowrap"
      >
        View
      </span>
    </button>
  </li>
  <li
    className="w-fit list-none  relative"
  >
    <button
      className="pl-2 cursor-pointer px-1 flex gap-1 sm:gap-2 items-center text-sm hover:bg-inactive-gray leading-5 py-1 hover:bg-active-gray rounded-sm justify-center w-fit  after:h-5 after:left-0 after:w-0.5 after:absolute after:bg-link-colour "
      onClick={[Function]}
      type="button"
    >
      <span
        className="whitespace-nowrap"
      >
        Fund
      </span>
    </button>
  </li>
</div>
`;
